src/common-libs/plan/index.ts

```diff
@@ -0,0 +1,511 @@
+  });
+  const latestTagName =
+    (latestPlanTag.length && latestPlanTag[0].name) || '0.0';
+  let [majorVersion, minorVersion] = latestTagName
+    .split('.')
+    .map((v) => parseInt(v) || 0);
+  if (planTag.major) {
+    majorVersion++;
+    minorVersion = 0;
+  } else {
+    minorVersion++;
+  }
+  return `${majorVersion}.${minorVersion}`;
+};
+
+export const setPlanReportingPeriod = async (
+  models: Database,
+  planTag: AddPlanTagInput
+) => {
+  const reportingPeriods = await models.planReportingPeriod.find({
+    where: {
+      planId: createBrandedValue(planTag.planId),
+    },
+    orderBy: { column: 'periodNumber', order: 'asc' },
+  });
+  if (!reportingPeriods.length) {
+    throw new Error(
+      'Cannot activate planTag if associated plan has no reporting period'
+    );
+  }
+  const latestPlanVersion = await models.planVersion.findOne({
+    where: {
+      planId: createBrandedValue(planTag.planId),
+      latestVersion: true,
+    },
+  });
+  if (latestPlanVersion && !latestPlanVersion.currentReportingPeriodId) {
+    await models.planVersion.update({
+      values: {
+        currentReportingPeriodId: reportingPeriods[0].id,
+      },
+      where: {
+        id: latestPlanVersion.id,
+      },
+    });
+  }
+};
+
+const getAssociatedPlanVersion = async (
+  models: Database,
+  planTag: InstanceDataOfModel<Database['planTag']>
+) => {
+  const associatedPlanVersion = await models.planVersion.findOne({
+    where: {
+      planId: planTag.planId,
+      createdAt: {
+        [models.Op.LTE]: planTag.createdAt,
+      },
+      latestVersion: true,
+    },
+  });
+  if (!associatedPlanVersion) {
+    throw new Error('Cannot activate planTag if no associated version exist');
+  }
+  return associatedPlanVersion;
+};
+
+export const updateLastPublishedReportingPeriodId = async (
+  models: Database,
+  planTag: InstanceDataOfModel<Database['planTag']>
+) => {
+  const reportingPeriods = await models.planReportingPeriod.find({
+    where: {
+      planId: planTag.planId,
+      measurementsGenerated: true,
+    },
+    orderBy: { column: 'periodNumber', order: 'desc' },
+  });
+
+  if (reportingPeriods.length) {
+    const latestReportingPeriodId = reportingPeriods[0].id;
+    const publishedMeasurements = await models.measurement.find({
+      where: {
+        planReportingPeriodId: latestReportingPeriodId,
+        versionTags: { [models.Op.NOT_IN]: [['{}']] },
+      },
+    });
+    console.log(publishedMeasurements.length);
+    if (publishedMeasurements.length) {
+      const associatedPlanVersion = await getAssociatedPlanVersion(
+        models,
+        planTag
+      );
+      console.log(associatedPlanVersion);
+      await models.planVersion.update({
+        values: {
+          lastPublishedReportingPeriodId: latestReportingPeriodId,
+        },
+        where: {
+          id: associatedPlanVersion.id,
+        },
+      });
+    }
+  }
+};
+
+const versionedModels: any = [
```

Comment:
Don't use `any`, except in special circumstances when defining type generics.

From this point forward, all the code should be moved to a separate file, because it is all related to versioning, so that way we can have everything isolated in a single place. Read my main comment to see the proposed contents of this file.